Neptune-Crypto/neptune-core
# add get_confirmations() RPC and add to dashboard

Addresses #6.  (partial)

The major objective of this PR is adding the `get_confirmations` RPC.   A secondary objective is displaying the confirmations field in overview screen of the Dashboard.   Other changes are ancillary.

Server:
* `implement get_confirmations` RPC
* return confirmations field in `get_dashboard_overview_data` RPC
* add `wallet_state::get_latest_balance_height()`
* add `block_height` to tuple returned from `wallet_state::get_balance_history`

Dashboard:
* display confirmations in overview screens
* change confirmations field from `usize` to `u64`

Notes:

* The `get_confirmations()` RPC function has been simplified and made more efficient than the prototype [here](https://github.com/Neptune-Crypto/neptune-core/issues/6#issuecomment-1732668472).
* `get_latest_balance_height()` will need to be modified to take into account unsynced/abandoned utxo.  There is already #28 for this, so I think it is best to do in a follow-up PR.  Also, I'm not yet certain of the correct fix, as I did try to simply filter out by `utxo.is_abandoned` in `get_balance_history()` and that did not solve the balance/history sync issue.   suggestions welcome.
* I considered returning additional fields from `get_latest_balance_height()` and just calling it `get_latest_balance()`.  However the amount field requires an additional call+loop and since there is no present need, I decided to keep it as simple/fast as possible.
* As a first pass at optimizing I had added `block_height` field to values returned from `wallet_state::get_balance_history()`.  That fn is no longer called from `get_confirmations()` however I left the `block_height` field in place because I think it would be useful to callers of the `get_history()` RPC.  I think it would be useful to display the height in Dashboard history, but decided to leave that change for a follow-up PR.  Anyway, this note explains that diff.
* I'm a bit uncertain yet about the correct type(s) for block heights (and counts).   There is the `BlockHeight` type which the compiler treats as `i128`.  But it doesn't really make sense for a height to be signed.  Also, `i128` and `u128` both seem like overkill... how many blocks are we anticipating?    Is `u64` not enough?    The dashboard was previously using `usize` for `confirmations` field, which seems a bit loose as it could be `u32` on some platform, so I changed it to `u64` and also my reasoning is less data over the wire vs `u128`.  But I would rather just be consistent with a single type (based on `u64`?) used everywhere to represent block heights.  Perhaps the RPC and Dashboard should all be using the `BlockHeight` for `confirmations` field rather than a primitive type.  thoughts?

Testing:

* Restarted server and Dashboard, then verified that confirmations are displayed in the overview screen to the right of `synced balance`
* Verified that confirmations field is zero when a new block is mined to my wallet, and increases by 1 for each subsequent block.
* Compared the `synced balance` with top-most balance in `History` screen.  The history screen balance is higher than `synced balance`.  This is wrong but expected until #28 is fixed.

I wanted to create a unit test for `get_confirmations` RPC however it is unclear to me how to do so.  There are a few unit tests in rpc_server.rs, but afaict none of them setup a wallet with a non-zero balance which would be a requirement for my test(s).  If someone could provide such a setup fn, I would happily make use of it.
